File: src/com/mcblox/parkour/Parkour.java
package com.mcblox.parkour;

import static com.mcblox.parkour.enums.ParkourSubCommands.*;

import org.bukkit.Bukkit;
import org.bukkit.entity.Player;
import org.bukkit.plugin.java.JavaPlugin;

import com.mcblox.parkour.cmds.CmdParkour;
import com.mcblox.parkour.enums.FolderStructure;
import com.mcblox.parkour.events.ParkourPlay;
import com.mcblox.parkour.objects.Course;
import com.mcblox.parkour.utils.Console;
import com.mcblox.parkour.utils.TitleAPI;

public class Parkour extends JavaPlugin {

	public static String VERSION;
	public static JavaPlugin PLUGIN;
	
	@Override
	public void onEnable() {
		
		// Init plugin static variables
		PLUGIN = this;
		VERSION = getDescription().getVersion();
		
		// Create File Structure
		for (FolderStructure folder : FolderStructure.values()) {
			folder.create();
		}

		// Register commands
		CmdParkour cmdParkour = new CmdParkour();
		PLUGIN.getCommand("parkour").setExecutor(cmdParkour);
		PLUGIN.getCommand("parkour").setTabCompleter(cmdParkour);

		// Register command events
		Bukkit.getPluginManager().registerEvents(SET_START.getCmd(), PLUGIN);
		Bukkit.getPluginManager().registerEvents(REGION.getCmd(), PLUGIN);
		Bukkit.getPluginManager().registerEvents(SET_FINISH.getCmd(), PLUGIN);
		Bukkit.getPluginManager().registerEvents(DELETE.getCmd(), PLUGIN);
		Bukkit.getPluginManager().registerEvents(ADD_CHECK.getCmd(), PLUGIN);
		
		// Register stand-alone events
		Bukkit.getPluginManager().registerEvents(new ParkourPlay(), PLUGIN);

		// Load Courses
		Course.loadCourses();
		
		// Done
		Console.info("Enabled!");
		
	}

	@Override
	public void onDisable() {
		
		// Clear Titles
		for (Player player : Bukkit.getOnlinePlayers()) {
			TitleAPI.sendTitle(player, 0, 0, 0, "", "");
		}
		
		// Save courses
		for (Course course : Course.courses) {
			course.saveCourse();
			course.show();
		}
		
		// Disabled
		Console.info("Disabled!");
		
	}
	
}

File: src/org/serversmc/parkour/core/Parkour.kt
package org.serversmc.parkour.core

import org.bukkit.*
import org.bukkit.plugin.java.*
import org.serversmc.parkour.cmds.*
import org.serversmc.parkour.cmds.pk.*
import org.serversmc.parkour.events.*
import org.serversmc.parkour.utils.*

lateinit var PLUGIN: Main

class Main : JavaPlugin() {
	
	override fun onEnable() {
		// Initialize libraries
		PLUGIN = this
		// Create Data folder
		if (!dataFolder.exists()) dataFolder.mkdirs()
		// Register commands
		registerCommands()
		// Register Events
		Bukkit.getPluginManager().apply {
			registerEvents(BlockBreak, PLUGIN)
			registerEvents(PlayerMove, PLUGIN)
			registerEvents(PlayerQuit, PLUGIN)
			registerEvents(SensorInteract, PLUGIN)
		}
		// Load Courses
		CourseManager.loadCourses()
		// Check for Update
		// TODO
		// Done
		Console.info("Loaded")
	}
	
	override fun onDisable() {
		// Save courses and show them
		CourseManager.saveAndClose()
		// Clear managers
		SelectManager.clear()
		EventTracker.clear()
		// Done
		Console.info("Done")
	}
	
	private fun registerCommands() {
		// Main commands
		CParkour.register()
		// Parkour sub commands
		CAddRegion.register()
		CCreate.register()
		CDelete.register()
		CHelp.register()
		CInfo.register()
		CList.register()
		CSelect.register()
		CSetFinish.register()
		CSetSpawn.register()
		CSetStart.register()
	}
	
}